plugins/txn_test_gen_plugin/txn_test_gen_plugin.cpp

```diff
@@ -306,29 +306,84 @@ struct txn_test_gen_plugin_impl {
 
    int32_t txn_reference_block_lag;
 
+   int startup_produce_blocks = 0;
+   private_key_type producer_key;
+
    abi_serializer eosio_token_serializer = fc::json::from_string(eosio_token_abi).as<abi_def>();
 };
 
 txn_test_gen_plugin::txn_test_gen_plugin() {}
 txn_test_gen_plugin::~txn_test_gen_plugin() {}
 
 void txn_test_gen_plugin::set_program_options(options_description&, options_description& cfg) {
+   auto default_priv_key = private_key_type::regenerate<fc::ecc::private_key_shim>(fc::sha256::hash(std::string("nathan")));
+   auto private_key_default = std::make_pair(default_priv_key.get_public_key(), default_priv_key );
    cfg.add_options()
       ("txn-reference-block-lag", bpo::value<int32_t>()->default_value(0), "Lag in number of blocks from the head block when selecting the reference block for transactions (-1 means Last Irreversible Block)")
+      ("startup-produce-blocks", bpo::value<int32_t>()->default_value(0), "number of block being produced at startup")
+      ("startup-sign-key", boost::program_options::value<vector<string>>()->composing()->multitoken()->default_value({fc::json::to_string(private_key_default)},
+                                                                                                fc::json::to_string(private_key_default)), "key to sign startup blocks")
    ;
 }
 
+template<typename T>
+T dejsonify(const string& s) {
+   return fc::json::from_string(s).as<T>();
+}
+
 void txn_test_gen_plugin::plugin_initialize(const variables_map& options) {
+   ilog("initialize txn_test_gen plugin");
+
    my.reset(new txn_test_gen_plugin_impl);
    my->txn_reference_block_lag = options.at("txn-reference-block-lag").as<int32_t>();
+
+   my->startup_produce_blocks = options.at("startup-produce-blocks").as<int32_t>();
+   if (options.count("startup-sign-key")) {
+      const std::vector<std::string> key_id_to_wif_pair_strings = options["startup-sign-key"].as<std::vector<std::string>>();
+      for (const std::string& key_id_to_wif_pair_string : key_id_to_wif_pair_strings)
+      {
+         auto key_id_to_wif_pair = dejsonify<std::pair<public_key_type, private_key_type>>(key_id_to_wif_pair_string);
+         my->producer_key = key_id_to_wif_pair.second;
+         break;
+      }
+   }
 }
 
 void txn_test_gen_plugin::plugin_startup() {
+   ilog("starting txn_test_gen plugin");
    app().get_plugin<http_plugin>().add_api({
       CALL(txn_test_gen, my, create_test_accounts, INVOKE_V_R_R(my, create_test_accounts, std::string, std::string), 200),
       CALL(txn_test_gen, my, stop_generation, INVOKE_V_V(my, stop_generation), 200),
       CALL(txn_test_gen, my, start_generation, INVOKE_V_R_R_R(my, start_generation, std::string, uint64_t, uint64_t), 200)
    });
+
+   if (my->startup_produce_blocks > 0) {
+      ilog("ready to generate ${i} initial block(s)", ("i", my->startup_produce_blocks));
```

Comment:
This method of initializing a chain will be pretty cumbersome.
- because of the new explicit nature of block start/commit/abort we've centralized all control of blocks in the `producer_plugin` which must track additional state to properly operate.  This direct access of the `controller` will desync that state
- this does not seem to detect whether the chain is new, so on replay/restart it will create an additional `startup_produce_blocks` worth of blocks, i suspect on many chains this will assert out and cause the `nodeos` process to stop
- This does not guarantee that between the timestamp in genesis.json (assuming a new chain) and `fc::time_point::now()` there is room for `startup_produce_blocks` worth of block time.  As a result, it is entirely possible that the block that is started at the end is in the "past" 

the second points are solvable.  The first point is a deal breaker of sorts.  We don't want debugging features to leak into the `producer_plugin` and unfortunately it has invariants to maintain.  We can pass signed blocks into it the same way the `net_plugin` does but you would need a way of generating these signed blocks without corrupting the state of chain controller.  Perhaps for empty blocks this may be doable.

Reply:
In my understanding these changes are just for testing purpose for ticket #2404 in which we need to test the sync-up between nodes of different chain lengths. And they're all in txn_test_gen_plugin. For point 3, I guess I can figure out how many blocks at maximum we can insert between the genesis timestamp and time_point::now(). But for point 1, would you suggest some better idea?

Reply:
unfortunately it is not possible to run a node without the producer_plugin at this time.  Even though the changes are isolated they are violating some of the (undocumented) ownership model that the plugins need to respect for node stability.

Reply:
If by this you mean, this feature should never be used with an actual producer (it will produce blocks in this plugin and no where else for the life of the process), it may work.  In that case, I would suggest isolating it in its own plugin.  The txn_test_gen_plugin _is_ used in testnets which have producers and leaving an option that will silently try to corrupt things in a plugin that gets use is probably a bad idea.

Reply:
Got it. I'll isolate it into a new testing plugin then. Thanks!